docs/decisions/0021-transaction-aggregates.rst

```diff
@@ -0,0 +1,135 @@
+   it won't help us keep the aggregate records up-to-date in a timely manner (we'd have to run
+   it on some cron/schedule).
+2. Updating the aggregates based on the ``redeem`` response body (on successful redemption).
+   This can help us update aggregates as redemptions occur, but won't help us backpopulate
+   aggregates.  Furthermore, it doesn't help us update aggregates based on any reversal records
+   that are committed.
+3. Emiting transaction/reversal commit events from the enterprise-subsidy service. If we can
+   emit these to an event broker, consumable by enterprise-access, then we can update aggregates
+   as transactions and reversals are committed. Furthermore, we could utilize a Django management
+   command to emit events that already exist (perhaps with ``modified`` times
+   earlier than some adjustable value) as a way to replay history. This would allow us to backpopulate
+   aggregate records.
+
+We'll utilize option (3) to both backpopulate and update aggregate records in real time.
+
+Ensuring Data Consistency
+-------------------------
+Timely data consistency is one of the main benefits of our ``can_redeem`` and ``redeem`` flows.
+The communication along both flows is:
+
+* **synchronous** - it occurs via blocking requests that are initiated from a single service (enterprise-access)
+* **transactional** - no transactions are committed without a valid enrollment/fulfillment
+  action taking place, so from a business workload perspective, a single flow execution is transactional.
+  Note that this does not imply that any given *request* is transactional in the ACID sense.
+* **eventually consistent** - in the ``redeem`` workflow, enrollment records can be written to some state
+  in our most downstream services, and as soon a response is available to the subsidy service, a
+  existing *created* transaction record will transition its state to *committed*. So this flow is technically
+  eventually consistent, but the consistency is generally very timely (e.g. the time it takes for an
+  enrollment request from enterprise-subsidy -> enrollment services to resolve into a successful response).
+  So the synchronous nature of the flow helps lead to timely consistency, even if it is technically
+  "eventual".
+
+If we replace the synchronous querying of the enterprise-subsidy service for aggregates, with
+database reads of aggregate tables populated by asynchronous events, we've introduced asynchronicity
+into our overall flow.  This introduces the risk of running ``can_redeem`` flows with transaction
+aggregates that are stale, i.e. individual transaction records exist that have not been
+taken into accounts in the aggregate records. Without mitigation, this risks that our overall
+``redeem`` flow becomes less consistent, and timely consistency is one of the key cross-functional
+requirements of learner credit redemption.
+
+One possible mitigation is to record the timestamp of the most recent transaction or reversal
```

Comment:
+1 solid idea.

Thinking out loud I wonder what other ways we could go about check summing this. This approach would ensure we are pointing at the same index but it doesn't ensure we've read all the records... I'm not sure, maybe the offset is enough.

Reply:
Yeah, it's a good point. I think there's one more thing we have to take into account, which is the absence of a needed aggregate for some `redeem` flow - if you query for aggregates on a given (policy, learner), and no record is in the DB, do we assume the metric is 0, or that we haven't yet computed the aggregate? I think we can hook into the `can-redeem` flow, plus maybe formalize a `zero aggregate` concept to mitigate this. I'll add a paragraph to the doc.